docs/devguide/hexformat.rst

```diff
@@ -0,0 +1,70 @@
+.. _hexformat:
+
+=================
+Firmware Hex File
+=================
+
+When MicroPython is built, the compiler produces an
+`Intel Hex <https://en.wikipedia.org/wiki/Intel_HEX>`_ file containing the
+MicroPython firmware.
+Additional data can then be added to this file to contain information about the
+MicroPython version, or the Python code to execute on start-up.
+
+The general memory layout used is:
+
+- ``0x00000000``: Start of MicroPython firmware - up to 248 KBs
+- ``0x0003e000``: Start of appended script (optional) - up to 8 Kbs
+- ``0x100010c0``: UICR customer[16] register, start of MicroPython information - 28 bytes
+    
+.. note::
+    If you append any data or modify the Intel Hex file, please ensure the
+    addresses of the data stored progress in incremental order.
+    If there is an address jump backwards DAPLink will fail to flash the file.
+
+Appended script format
+----------------------
+
+MicroPython checks the first 2 bytes at address ``0x0003e000`` for a magic
+string to indicate if there is an appended script. If the magic string is
+found, it will automatically execute the Python code stored there, unless there
+is a main.py file stored in the MicroPython filesystem.
+
+- ``0x0003e000``: 2 bytes "MP"
+- ``0x0003e002``: 2 bytes, little endian integer for the length (in bytes) of the appended script (not counting this 4 byte header)
+- ``0x0003e004``: Script stored as bytes, for MicroPython to decode using utf-8.
+
+UICR format
+-----------
+
+The User Information Configuration Registers (UICR) is a region of Non-Volatile
+Memory available to store user-specific settings.
+The first 128 Bytes are reserved, but we can use the other 128 Bytes to store
+any arbitrary data.
+
+MicroPython stores the following information, in little endian, starting from
+the UICR customer[16] register:
+
+- ``0x100010c0``: 4-byte integer with magic value ``0x17eeb07c``
+- ``0x100010c4``: 4-byte integer with value ``0xffffffff``
+- ``0x100010c8``: 4-byte integer with value ``0x0000000a`` (log base 2 of the flash page size, being 1024 bytes)
+- ``0x100010ca``: 2-byte integer with value ``0x0000`` (start page of the firmware)
+- ``0x100010cc``: 2-byte integer storing number of pages used by the firmware
+- ``0x100010d0``: 4-byte integer with value ``0xffffffff``
+- ``0x100010d4``: 4-byte integer with the address in the firmware of the version string
+- ``0x100010d8``: 4-byte integer with value ``0x00000000``
+
+Steps to create the firmware.hex file
+-------------------------------------
+
+The yotta tool is used to build MicroPython, but before that takes place
+additional files have to be generated by the Makefile in preparation for the 
+build, and additional data is added to the hex file after.
+
+Running the ``make all`` command executes the following steps:
+
+- Interned strings are added to the ``qstrdefsport.h`` file
```

Comment:
This needs to be done manually, it's not done by `make`.  Probably best to just remove it.

Reply:
👍 I'll remove

Reply:
I do have a personal aversion to calling it firmware (as a firmware engineer, this is not firmware, it is a runtime image and library packaged alongside the user application). 

Might be too late to fix now, but we do get confusion occuring when we say 'update the firmware' and we mean 'update the interface chip firmware by putting the micro:bit into MAINTENANCE mode' and they think they have to update the MicroPython firmware.

While it does pass some of the early 'what is firmware' tests in this article (nvol, device access etc), the strict difference here is that firmware is mostly programmed once and only updated a small number of times across the lifecycle of a product.

https://en.wikipedia.org/wiki/Firmware

It's a personal opinion I have, feel free to ignore me, but I felt this was the best time to mention it if you are updating docs.

Reply:
I completely understand the confusion it can create since we call DAPLink the "microbit firmware". I'm happy to call it something else, but I'm not 100% sure what's the best term. Is it "MicroPython runtime"? @dpgeorge do you have any other suggestions?

Reply:
I think it's too late now to change these names.  With the latest Mu it is anyway more like firmware because there only the user scripts are updated on the filesystem, not the firmware anymore.

Reply:
The issue with changing it now is that it requires changes to the Makefile, at least.  The word "firmware" is referenced a lot in the code base (`git grep firmware` shows 40 lines).

Reply:
> With the latest Mu it is anyway more like firmware because there only the user scripts are updated on the filesystem, not the firmware anymore.

Users will still flash the full flash when switchin to MakeCode or the online Python editor, so that is still a relatively common action.

I wouldn't change any actual code or things like the Makefile, but there is a lot of documentation that will have to be updated (https://github.com/bbcmicrobit/micropython/issues?q=is%3Aopen+is%3Aissue+label%3Adocs) and **if** we decide to change it, then we could edit the occurrences we find them.

As a small note, it ooks like "MicroPython runtime" has only been used once, in the storage tutorial: https://bbcmicrobitmicropython.readthedocs.io/en/latest/tutorials/storage.html

To be completely honest I don't feel strongly either way.